### PR TITLE
feat: add usePagination hook

### PR DESCRIPTION
**Pull request by petermakowski**
## Done

- add `usePagination` hook (source: https://github.com/canonical-web-and-design/maas-ui/pull/3684#discussion_r833927576)
- refactor MainTable to use the `usePagination` hook

### Storybook

To see rendered examples of all react-components, run:

```shell
yarn start
```

### QA in your project

from `react-components` run:

```shell
yarn build
npm pack
```

Install the resulting tarball in your project with:

```shell
yarn add <path-to-tarball>
```

### QA steps

- Steps for QA.

## Fixes

Fixes: canonical-web-and-design/app-tribe#772

### CODE REVIEW

File: src/components/MainTable/MainTable.tsx [hunk 1]
```diff
@@ -316,6 +307,27 @@ const MainTable = ({
...
+    [currentSortDirection, currentSortKey, rows, sortable, sortFunction]
+  );
+
+  const {
+    pageData: finalRows,
+    currentPage,
+    paginate: setCurrentPage,
+  } = usePagination(sortedRows, {
+    itemsPerPage: paginate,
+    autoResetPage: true,
```
Comment (huwshimi) [inline src/components/MainTable/MainTable.tsx:328]: Should this be exposed as a component prop?

File: src/hooks/usePagination.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { useMemo, useEffect, useState } from "react";
+
+/**
+ * A hook that handles pagination.
+ * @param data The data array to paginate.
+ * @param itemsPerPage Number of items per page. Returns all items if no value has been provided.
+ * @param initialPage?: Initial page number. Defaults to 1.
+ * @param autoResetPage Whether to reset the page number to 1 when the data changes.
```
Comment (huwshimi) [inline src/hooks/usePagination.ts:8]: These need to be described inside `options`.

File: src/hooks/usePagination.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { useMemo, useEffect, useState } from "react";
+
+/**
+ * A hook that handles pagination.
+ * @param data The data array to paginate.
+ * @param itemsPerPage Number of items per page. Returns all items if no value has been provided.
+ * @param initialPage?: Initial page number. Defaults to 1.
```
Comment (huwshimi) [inline src/hooks/usePagination.ts:7]: Is `?:` the right syntax? Is the type hint enough to show that this is optional?

File: src/hooks/usePagination.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  const [pageIndex, setPageIndex] = useState(initialPageIndex);
+  const startIndex =
+    typeof itemsPerPage === "number" ? pageIndex * itemsPerPage : 0;
+  const paginate = (pageNumber: number) => setPageIndex(pageNumber - 1);
+
+  useEffect(() => {
+    if (typeof itemsPerPage === "number" && startIndex >= totalItems) {
+      !autoResetPage && Math.floor(totalItems / itemsPerPage) > 0
+        ? // go to the last available page if the current page is out of bounds
+          setPageIndex(Math.floor(totalItems / itemsPerPage) - 1)
```
Comment (huwshimi) [inline src/hooks/usePagination.ts:44]: This could do with a test.

File: src/hooks/usePagination.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        : // go to the initial page if autoResetPage is true
+          setPageIndex(0);
```
Comment (huwshimi) [inline src/hooks/usePagination.ts:46]: This could also do with a test.

File: src/components/MainTable/MainTable.tsx [hunk 6]
```diff
@@ -190,60 +191,20 @@ const generateHeaders = (
...
-    }
-    slicedRows = sortedRows.slice(startIndex, startIndex + paginate);
-  }
-  const rowItems = slicedRows.map(
+const generateRows = ({
+  rows,
+  headers,
+  responsive,
+  expanding,
+}: Partial<Props> & {
```
Comment (huwshimi) [inline src/components/MainTable/MainTable.tsx:199]: I don't think `Partial<Props>` is the right type here as this function requires a specific subset of the props.

**Review by huwshimi — APPROVED** [review]
Looking good @petermakowski. A few small things and some tests to add.

It would be good to follow up from this and add pagination to `ModularTable`.

File: src/components/MainTable/MainTable.tsx (hunk 1 above)
Comment (petermakowski) [reply src/components/MainTable/MainTable.tsx:330]: Not at this stage, I don't think. We can always add this if needed.

**Pull request merged**